Np 1250 create emscripten binding

saumyaj3 opened the pull request:
NP-1250

saumyaj3 committed:
Implement Emscripten support with JavaScript bindings for `UvulaJS`

- Added support for building JavaScript bindings through Emscripten.
- Introduced `WITH_JS_BINDINGS` option in CMake for enabling/disabling JS bindings.
- Updated `conanfile.py` to include new options, dependencies, and packaging logic for `UvulaJS`.
- Integrated `UvulaJS` bindings for import/export with TypeScript support.
- Modified GitHub actions to support WebAssembly (`platform_wasm`) and npm package workflows.

NP-1250

saumyaj3 committed:
Update GitHub workflow naming: `conan-package.yml` → `package.yml`

NP-1250

saumyaj3 committed:
Fix workflow path trigger in package.yml to include correct location

NP-1250

casperlamboo commented on UvulaJS/UvulaJS.cpp:
```diff
@@ -0,0 +1,437 @@
+
+uvula_info_t get_uvula_info()
+{
+    return { UVULA_VERSION, "" }; // Git hash can be added later if needed
```
Can we just remove the hash if we're not adding it?

casperlamboo reviewed (changes requested):
For both the `unwrapTyped` and `projectTyped` function I have two similar remarks. They both return a `emscripten::val` which is _untyped_ when integrated in a host platform. This seems counter to the comments around. Additionally I see you use `std::vector<T>` as arguments. Using cpp vectors in typescript is hella annoying. In neoprep we've often opted to just accept javascript lists and perform the conversion in cpp.

saumyaj3 committed:
Refactor UvulaJS JavaScript bindings for TypeScript integration

- Replaced custom structures with standardized TypeScript-like types (Float32Array, Uint32Array, Int32Array).
- Refactored wrappers and return types to improve type safety in TypeScript.
- Enhanced `project` and `unwrap` methods with new structured return types.
- Updated Emscripten bindings to support TypeScript-friendly structures and arrays.

NP-1250

saumyaj3 committed:
Expand GitHub Actions path triggers in `package.yml` to include `UvulaJS` directory

NP-1250

casperlamboo commented on UvulaJS/UvulaJS.cpp:
```diff
@@ -378,10 +421,23 @@ emscripten::val jsProject(
-    register_vector<uint32_t>("UInt32Array");
+    // Register typed array types
+    register_vector<float>("Float32Array");
+    register_vector<std::vector<float>>("Float32ArrayArray");
```
There are not needed right?

```suggestion
```

saumyaj3 committed:
Refactor `UvulaJS` bindings to enhance TypeScript compatibility

- Introduced `PolygonArray` class to wrap arrays for improved TypeScript typing.
- Replaced `std::vector` with `Float32Array` and `PolygonArray` in key data structures.
- Enhanced `unwrap` and `project` methods to utilize TypeScript-friendly return types.
- Updated Emscripten bindings to register `PolygonArray` and support structured data types.

NP-1250

saumyaj3 committed:
Add Emscripten-specific threading configuration and refactor memory allocation

- Disabled multithreading for Emscripten/WASM builds.
- Simplified `TaskGroup` allocation logic.

NP-1250

casperlamboo commented on UvulaJS/UvulaJS.cpp:
```diff
@@ -0,0 +1,517 @@
+        params.cameraNormal,
+        params.faceId
+    );
+}
```
Why do we need this indirection function? Why can't we directly expose the `projectTyped` function?

casperlamboo commented on UvulaJS/UvulaJS.cpp:
```diff
@@ -0,0 +1,517 @@
+    uint32_t face_id)
+{
+    // Convert stroke polygon (flat array of [x1, y1, x2, y2, ...])
+    std::vector<Point2F> stroke_points;
+    unsigned stroke_length = stroke_polygon_js["length"].as<unsigned>();
+    stroke_points.reserve(stroke_length / 2);
+    for (unsigned i = 0; i < stroke_length; i += 2) {
+        float x = stroke_polygon_js[i].as<float>();
+        float y = stroke_polygon_js[i + 1].as<float>();
+        stroke_points.push_back({x, y});
+    }
```
Untested, but I think we can prevent copying the whole geometry here by using code similar to this.
```suggestion
    const std::span<Point2F> stroke_polygon = std::span(reinterpret_cast<Point2F*>(stroke_polygon_js.as<uintptr_t>()), stroke_polygon_js["length"].as<size_t>() / 2);
```

(code is untested!!!!) I would suggest implementing this at the end, once we know everything works. And then see if we can make this optimisation.

casperlamboo reviewed (commented):
Was working with libUvula bindings and found these remarks

casperlamboo committed:
Simplify bindings

NP-1237

casperlamboo committed:
Transpose camera matrix

saumyaj3 committed:
Merge branch 'main' into NP-1250-create-emscripten-binding-II

casperlamboo committed:
Merge remote-tracking branch 'origin/NP-1250-create-emscripten-binding-II' into NP-1250-create-emscripten-binding-II

casperlamboo committed:
Merge pull request #6 from Ultimaker/NP-1250-create-emscripten-binding-II

Simplify bindings

casperlamboo committed:
Construct `Float32Array` to return uv coords

NP-1250

casperlamboo committed:
Move vertices in `Geometry` container

NP-1250

casperlamboo committed:
Merge pull request #8 from Ultimaker/NP-1250-create-emscripten-binding-III

Create emscripten bindings

casperlamboo committed:
Fix js bindings so we can provide a `const Geometry&` in `project` function

NP-1250

casperlamboo committed:
Merge branch 'refs/heads/NP-1250-create-emscripten-binding-III' into NP-1250-create-emscripten-binding

wawanbreton committed:
Fix Python binding

NP-1250
